File: lib/ruby_impl/speech_trimmer.rb
module Noyes
  # SpeechTrimmer trims non-speech from both ends of an audio stream. Each time
  # you enqueue audio into it you should dequeue audio out of it until dequeue
  # returns nil.  Then check for eos.  If eos is true you are done.
  # SpeechTrimmer is designed to work efficiently with live audio.
  class SpeechTrimmer
    def initialize
      @leader = 5  # Cents of leading silence to retain.
      @trailer = 5  # Cents of trailing silence to retain.
      @speech_started = false
      @cent_marker = BentCentMarker.new
      @false_count=0
      @true_count=0
      @queue = []
      @eos_reached = false
      @scs = 20 # Centiseconds of speech before detection of utterance.
      @ecs = 50 # Centiseconds of silence before end detection.
    end

    def enqueue pcm
      return if @eos_reached
      @queue << pcm
      if @cent_marker << pcm
        @false_count = 0
        @true_count += 1
      else
        @false_count += 1
        @true_count = 0
      end
      if @speech_started
        if @false_count == @ecs
          @eos_reached = true
          # only keep trailer number of cents once eos is detected.
          @queue = @queue[0, @queue.size - @ecs + @trailer]
        end
      elsif @true_count > @scs
        # Discard most begining silence, keeping just a tad.
        if @leader + @scs < @queue.size
          start = @queue.size - @leader - 1 - @scs
          @queue = @queue[start,@queue.size - start]
        end
        @speech_started = true
      end
    end
    def dequeue
      if @eos_reached || (@speech_started && @queue.size > @ecs)
        @queue.shift
      end
    end
    def eos?
      @eos_reached
    end
  end    
end

File: lib/java_impl/talkhouse/SpeechTrimmer.java
package talkhouse;
import java.util.LinkedList;
import java.util.List;
import java.util.ArrayList;

public class SpeechTrimmer {
  int leader = 5;
  int trailer = 5;
  boolean speechStarted = false;
  BentCentMarker centMarker = new BentCentMarker(); 
  int falseCount=0;
  int trueCount=0;
  LinkedList<double[]> queue = new LinkedList<double[]>();
  boolean eosReached = false;
  int scs = 20;
  int ecs = 50;
  Segmenter seg;

  public SpeechTrimmer(int frequency) {
    seg = new Segmenter(frequency/100, frequency/100);
  }

  public double[][] apply(double[] pcm) {
    if (eos())
      return null;

    double[][] segments = seg.apply(pcm);
    if (segments == null)
      return null;

    List<double[]> resultQ = new ArrayList<double[]>(segments.length);
    for (int i=0;i<segments.length;++i) {
      enqueue(segments[i]);
      double[] centispeech = dequeue();
      while (centispeech != null) {
        resultQ.add(centispeech);
        centispeech = dequeue();
      }
      if (eos())
        break;
    }

    if (eos() && resultQ.size() == 0)
      return null;

    return resultQ.toArray(new double[resultQ.size()][]);
  }

  public void enqueue(double[] pcm) {
    if (eosReached)
      return;

    queue.offer(pcm);
    if (centMarker.apply(pcm)) {
      falseCount = 0;
      trueCount += 1;
    } else {
      falseCount += 1;
      trueCount = 0;
    }
    if (speechStarted) {
        if (falseCount == ecs) {
            eosReached = true;
            int newSize = queue.size() - ecs + trailer;
            queue = new LinkedList<double[]>(queue.subList(0, newSize));
        }
    } else if (trueCount > scs) {
      if (leader + scs < queue.size()) {
        int start = queue.size() - leader - scs - 1;
        int remainder = queue.size() - start;
        queue = new LinkedList<double[]>(queue.subList(
                                          start, start + remainder));
      }
      speechStarted = true;
    }
  }

  public double[] dequeue() {
    if (eosReached  && queue.size() > 0 || speechStarted && queue.size() > ecs) {
      return queue.remove();
    } else
      return null;
  }
  public boolean eos() {
    return eosReached;
  }
}
